### PR TITLE
First version of IGT-Splits

### PR DESCRIPTION
**Pull request by LukeSaward1**
Not going to close the issue yet since it's not fully complete.

### CODE REVIEW

File: tool/speedrun.re [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    fn format_time(self, time: float) -> float {
+        float::floor(time, 2)
+    }
+}
+
+static SPEEDRUN_COMPONENT = Component {
+    id: SPEEDRUN_COMPONENT_ID,
+    conflicts_with: List::of(SPEEDRUN_COMPONENT_ID),
+    draw_hud: fn(text: string) -> string {
+        let mut time = f"{SPEEDRUN_STATE.cur_time.to_int()/60}:{SPEEDRUN_STATE.cur_time.to_int() % 60:02}.{(SPEEDRUN_STATE.cur_time - SPEEDRUN_STATE.cur_time.to_int().to_float()) * 100.:02.0}";
```
Comment (oberien) [inline tool/speedrun.re:30]: This is quite a long line with lots of repetition. You should extract `SPEEDRUN_STATE.cur_time` to a variable.

File: tool/speedrun.re [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        ls.start_seconds.to_float() + ls.start_partial_seconds
+    }
+    fn start_timer(self) {
+        SPEEDRUN_STATE.timer_active = false;
+        SPEEDRUN_STATE.cur_time = 0.;
+        SPEEDRUN_STATE.timer_active = true;
+    }
+    fn format_time(self, time: float) -> float {
+        float::floor(time, 2)
+    }
```
Comment (oberien) [inline tool/speedrun.re:23]: unused function

File: tool/speedrun.re [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        let mut text = f"{time}\n{text}";
+        text
+    },
+    tick_mode: TickMode::DontCare,
+    requested_delta_time: Option::None,
+    on_tick: fn() {
+        match SPEEDRUN_STATE.timer_active {
+            true => SPEEDRUN_STATE.cur_time = float::floor(Tas::get_accurate_real_time() - SPEEDRUN_STATE.get_start_time(), 2),
+            false => SPEEDRUN_STATE.cur_time = SPEEDRUN_STATE.cur_time,
+        }
```
Comment (oberien) [inline tool/speedrun.re:40]: 1. noop `false` branch
2. no need to floor the value here, just floor it when displaying (which the format string is already doing)
3. if-else is usually better than matching a bool

```rust
if SPEEDRUN_STATE.is_timer_active {
    Tas::get_accurate_real_time() - SPEEDRUN_STATE.get_start_time()
}
```

File: tool/speedrun.re [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    },
+    on_yield: fn() {},
+    on_new_game: fn() {},
+    on_level_change: fn(old: int, new: int) {
+        match new {
+            31 => SPEEDRUN_STATE.timer_active = false,
+            0 => SPEEDRUN_STATE.start_timer(),
```
Comment (oberien) [inline tool/speedrun.re:47]: I'd switch those two branches around to make the numbers ascending.

File: tool/tas.re [hunk 5]
```diff
@@ -209,4 +209,4 @@ static TAS_COMPONENT = Component {
         TAS_STATE.events.push(InputEvent::MouseMoved(x, y));
     },
     on_component_exit: fn() {},
-};
+};
```
Comment (oberien) [inline tool/tas.re:212]: questionable ;)

**Review by oberien — CHANGES_REQUESTED** [review]
Just a few pedantic style changes. All in all pretty nice code! :+1:

File: tool/misc.re [hunk 6]
```diff
@@ -67,6 +67,14 @@ fn create_replay_menu(op: ReplayMenuOp) -> Ui{
...
 fn create_misc_menu() -> Ui {
     Ui::new("Misc:", List::of(
+        UiElement::Button(UiButton {
+            label: Text { text: "Speedrun" },
+            onclick: fn(label: Text) {
+                add_component(SPEEDRUN_COMPONENT);
+                leave_ui();
+                leave_ui();
+            }
+        }),
```
[the remark this answers is not in the document]
Comment (LukeSaward1) [reply tool/misc.re:77]: Need to add that.

File: tool/tas.re (hunk 5 above)
Comment (oberien) [inline tool/tas.re:212]: questionable ;)

File: tool/timer.re [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        TIMER_STATE.is_timer_active = true;
+    }
+}
+
+static TIMER_COMPONENT = Component {
+    id: TIMER_COMPONENT_ID,
+    conflicts_with: List::of(TIMER_COMPONENT_ID),
+    draw_hud: fn(text: string) -> string {
+        let mut foo = TIMER_STATE.cur_time;
+        let mut time = f"{foo.to_int()/60}:{foo.to_int() % 60:02}.{(float::floor(foo - foo.to_int().to_float(), 2)) * 100.:02.0}";
```
Comment (oberien) [inline tool/timer.re:27]: As discussed: `f"{time.to_int()/60}:{time.to_int() % 60:02}.{float::to_int(time * 100.) % 100:02}"`

File: tool/timer.re [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            TIMER_STATE.cur_time = Tas::get_accurate_real_time() - TIMER_STATE.get_start_time();
+        }
+    },
+    on_yield: fn() {},
+    on_new_game: fn() {},
+    on_level_change: fn(old: int, new: int) {
+        match new {
+            0 => TIMER_STATE.start_timer(),
+            31 => TIMER_STATE.is_timer_active = false,
+            _ => print(f"NEW: {new} | OLD: {old}"),
```
Comment (oberien) [inline tool/timer.re:44]: debug print should be removed

**Review by oberien — CHANGES_REQUESTED** [review]

**Review by oberien — APPROVED** [review]

**Pull request merged**